libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java

```diff
@@ -0,0 +1,222 @@
+      ((OffRouteDetector) offRoute).clearDistancesAwayFromManeuver();
+    }
+  }
+
+  /**
+   * If the {@link OffRouteCallback#onShouldIncreaseIndex()} has been called by the
+   * {@link com.mapbox.services.android.navigation.v5.offroute.OffRouteDetector}, shouldIncreaseStepIndex
+   * will be true and the {@link NavigationIndices} step index needs to be increased by one.
+   *
+   * @param navigation to get the next {@link LegStep#geometry()} and off-route engine
+   */
+  void checkIncreaseStepIndex(MapboxNavigation navigation) {
+    if (shouldIncreaseStepIndex) {
+      advanceStepIndex(navigation);
+      shouldIncreaseStepIndex = false;
+    }
+  }
+
+  Location buildSnappedLocation(MapboxNavigation mapboxNavigation, boolean snapToRouteEnabled,
+                                Location rawLocation, RouteProgress routeProgress, boolean userOffRoute) {
+    final Location location;
+    if (!userOffRoute && snapToRouteEnabled) {
+      location = getSnappedLocation(mapboxNavigation, rawLocation, routeProgress, stepPoints);
+    } else {
+      location = rawLocation;
+    }
+    return location;
+  }
+
+  /**
+   * Increases the step index in {@link NavigationIndices} by 1.
+   * <p>
+   * Decodes the step points for the new step and clears the distances from
+   * maneuver stack, as the maneuver has now changed.
+   *
+   * @param mapboxNavigation to get the next {@link LegStep#geometry()} and {@link OffRoute}
+   */
+  private void advanceStepIndex(MapboxNavigation mapboxNavigation) {
+    // First increase the indices and then update the majority of information for the new routeProgress
+    indices = increaseIndex(previousRouteProgress, indices);
+
+    // First increase the indices and then update the majority of information for the new
+    stepPoints = decodeStepPoints(mapboxNavigation.getRoute(), indices.legIndex(), indices.stepIndex());
+    updateOffRouteDetectorStepPoints(mapboxNavigation.getOffRouteEngine(), stepPoints);
+  }
+
+  /**
+   * Given the current {@link DirectionsRoute} and leg / step index,
+   * return a list of {@link Point} representing the current step.
+   * <p>
+   * This method is only used on a per-step basis as {@link PolylineUtils#decode(String, int)}
+   * can be a heavy operation based on the length of the step.
+   *
+   * @param directionsRoute for list of steps
+   * @param legIndex        to get current step list
+   * @param stepIndex       to get current step
+   * @return list of {@link Point} representing the current step
+   */
+  private List<Point> decodeStepPoints(DirectionsRoute directionsRoute, int legIndex, int stepIndex) {
+    // Check for valid legs
+    List<RouteLeg> legs = directionsRoute.legs();
+    if (legs == null || legs.isEmpty()) {
+      return stepPoints;
+    }
+    // Check for valid steps
+    List<LegStep> steps = legs.get(legIndex).steps();
+    if (steps == null || steps.isEmpty() || stepIndex >= steps.size()) {
```

Comment:
What about [decomposing conditional](https://refactoring.com/catalog/decomposeConditional.html) extracting that complicated expression into a well-named `private` method? Conditionals tend to get more and more complicated in their logic over time.